Hyperliquid

gaardiolor opened the pull request:
<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)
-->
## Summary

Hyperliquid: implemented spot, swap (long, short, leverage, stoploss_on_exchange)

Solve the issue: #10377

## Quick changelog

- Hyperliquid: implemented spot, swap (long, short, leverage, stoploss_on_exchange)

## What's new?
Hyperliquid: implemented spot, swap (long, short, leverage, stoploss_on_exchange)

Commit:
Initial implementation of hyperliquid.
- Spot
- Swap (long, short, leverage, stoploss_on_exchange)
- dry_run_liquidation_price()

Commit:
updated dry_run_liquidation_price to match new signature
fixed comment

xmatthias commented on requirements.txt:
```diff
@@ -4,7 +4,7 @@ bottleneck==1.4.2
 numexpr==2.10.1
 pandas-ta==0.3.14b
 
-ccxt==4.4.20
```
that's a pointless change and will cause a conflict tomorrow - and from there on every week the PR is open.

best leave version bumps to dependabot - unless there's a specific "Min version" requirement - where still - this would be the wrong place to bump.

gaardiolor replied:
sorry, new to all this. reverted

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -15,21 +19,145 @@ class Hyperliquid(Exchange):
+                "Freqtrade only supports isolated futures for leverage trading"
+            )
+
+    def get_funding_fees(
```
You can probably copy the test for this from `test_bybit` ...

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -15,21 +19,145 @@ class Hyperliquid(Exchange):
+            # Don't use set_leverage(), as this sets margin back to cross
+            self.set_margin_mode(pair, self.margin_mode, params={"leverage": leverage})
+
+    def dry_run_liquidation_price(
```
This will for sure need a dedicated test - ideally a parametrized one testing multiple scenarios (long,short, ...).

ideally, the exchange docs have a calculation example you can "copy" from (one that contains the necessary values and expected results) - which should be one of the tests.

gaardiolor replied:
ok! I followed the docs, so I think that's good. Tested it against ~170 unique ccxt `fetch_positions()` outputs on real orders, I'll make a test for that.

gaardiolor replied:
some info about why I chose those positions:
- different prices (BTC is high, DOGE is low, SOL & ETH in between)
- different sides
- different leverages (1-5)

max 1 position per symbol/side/leverage combination. Could remove some if you think this is overkill.

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -15,21 +19,145 @@ class Hyperliquid(Exchange):
+        self, timeframe: str, candle_type: CandleType, since_ms: Optional[int] = None
+    ) -> int:
+        # Funding rate candles have a different limit
+        if candle_type in CandleType.FUNDING_RATE:
```
```suggestion
        if candle_type == CandleType.FUNDING_RATE:
```

xmatthias reviewed (commented):
some preliminary, early feedback...

It'll surely need tests for the 2 "big" override functions (see explicit comments) - especially for the dry-liquidation calculation one.

Ci will fail currently (which i can see from the code) - if you follow the [contributing guide](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md#before-sending-the-pr) (deeplink to the relevant section) - you should most likely be able to use ruff's auto-fix to fix most errors (though i didn't check if it can fix all for now).

gaardiolor committed:
fix ohlcv_candle_limit funding_rate candles

Co-authored-by: Matthias <xmatthias@outlook.com>

gaardiolor committed:
revert ccxt version bump

Commit:
fixes related to ruff and mypy

gaardiolor commented:
ok, I fixed the ruff and mypy things. reverted the ccxt bump, hope that fixes it. Committed all that. 

Need some guidance on the following. Just say if you want me to commit everything so you can get a better understanding (not sure what you prefer).

## Testing environment
I had to do a couple of changes to files other than test_hyperliquid.py to make pytest work in my environment. Is this ok to commit?

### conftest.py
To fix `RuntimeError: aiodns needs a SelectorEventLoop on Windows`
```
def pytest_sessionstart(session):
    if sys.platform.startswith("win"):
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
```

### test_ccxt_compat.py
To fix `TypeError: Converting from datetime64[ns, UTC] to int32 is not supported. Do obj.astype('int64').astype(dtype) instead` . To fix:
```
< assert exch.klines(pair_tf)["date"].astype(int).iloc[0] // 1e6 == since_ms
> assert exch.klines(pair_tf)["date"].astype("int64").iloc[0] // 1e6 == since_ms
```

## Tests
Copied the funding_fees test from bybit, works.

Made a test for `dry_run_liquidation_price()`, also works. 
* placed markets data in tests/testdata/hyperliquid/markets.json (to get the max leverage)
* placed position data in tests/testdata/hyperliquid/positions.json
* reading those with @pytest.fixture and passing them to the test
* overwriting exchange.markets (`dry_run_liquidation_price()` needs them)
* the tests check if the calculated liq price ~matches the liq price returned by ccxt `fetch_positions()`

Is that a good approach? I don't see any of the other test_<exchange>.py files read a json. Didn't want to put them in the testfile itself, they're pretty big.. markets.json is 326 lines, positions.json 1394 lines.

Again, I can also just commit everything if that's easier, let me know.. thx

Commit:
typo

xmatthias commented:
your local test problems - no.
These are due to a wrong / faulty setup (CI on windows runs fine - so it's not a freqtrade / test issue - please fix your environment).
The 2nd issue might point to a wrong pandas or numpy version ... please make sure to have the ones installed we're pinning in the requirements files.


As for reading json files - you **can** do that - but we usually don't as it's not necessary. 
We won't need the full hyperliquid response / markets - but only the few we're actually testing against.

i doubt 2 markets cause 300 lines of json - and the test-data isn't supposed to be complete - only sufficient for what the test requires - we'll not be updating it just because an exchange adds a new market.

For positions i'm not sure - is that your trade positions ? 
then i'd reduce it to 2-3 positions (make sure to remove / alter identifying information before committing).


I'd say - commit the ones for hyperliquid and we'll see.

Commit:
check for tickers_have_bid_ask

Commit:
added hyperliquid to conftest

Commit:
added tests for hyperliquid

gaardiolor commented:
> The 2nd issue might point to a wrong pandas or numpy version ... please make sure to have the ones installed we're pinning in the requirements files.

Yeah, I've got pandas==2.2.3 numpy==1.26.4 like in requirements.txt . numpy==2.1.2 doesn't give the error. However, in the freqtrade container it works fine with numpy==1.26.4 , so it's a bit strange. int behaviour seems to be different, despite the versions being the same. The container defaults to int64, my environment defaults to int32. 

```
import sys
import pandas as pd
import numpy as np
print(f"python:{sys.version} pandas:{pd.__version__} numpy:{np.__version__} result:{pd.Series(1).astype(int).dtype}")
```
container: `python:3.12.7 (main, Oct 19 2024, 01:09:09) [GCC 12.2.0] pandas:2.2.3 numpy:1.26.4 result:int64`
a test venv on my system: `python:3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:17:27) [MSC v.1929 64 bit (AMD64)] pandas:2.2.3 numpy:1.26.4 result:int32`

But, I guess you have better things to do than troubleshooting my environment :)

I have added the tests. I already trimmed down the markets to only the 4 I need, but further trimmed them to only include the key I need. Same for positions. So the data is in `test_hyperliquid.py` just like with the other tests.

gaardiolor commented:
Hi @xmatthias , just checking in to make sure you're not waiting for me at this moment. Let me know if I can do anything. thx

xmatthias commented on freqtrade/exchange/exchange.py:
```diff
@@ -1274,6 +1275,8 @@ def create_order(
                 params,
             )
+            if not self._ft_has.get("create_order_has_all_data"):
+                order = self._api.fetch_order(order["id"], pair)
```
I'm not sure this is a good approach - as it could also lead to additional (timing) problems.
Depending on the exchange, it may take a few moments for the just placed order to actually appear in other endpoints. By doing this call "right away" - you risk getting a "order not found" exception here (even though we've placed it correctly) - which would lead to the bot thinking the placement never went through.

A potentially better approach might be how it's handled for kucoin - which simply sets type and status to the order response (could be moved to this location...? ) - and leaves the actual fetching of the order to the code designed to actually do that.

gaardiolor replied:
ok, the initial issue I noticed was that, when entering an order, freqtrade logged (and sent telegram messages) that stated `amount: 0`. Traced that back to that information not being available in the `create_order()` return info of ccxt. Then saw that the hyperliquid api doesn't return it as well (so it can't be fixed in ccxt). 

How does the kucoin implementation deal with that.. I would expect it to to log / send telegram messages with `amount: 0` ?

I was also worried about timing issues. But, what I found it that hyperliquid always (well, I tested with about 100 positions) returns that info when doing fetch_order directly after creating it. So it looks to be reliable. 

I could move that fetch_order logic to the hyperliquid class in a similar manner to kucoin (see below). Pro's is that it doesn't introduce another config variable. Cons is that once another exchange needs it, that config variable might be cleaner.

```
    def create_order(
        self,
        *,
        pair: str,
        ordertype: str,
        side: BuySell,
        amount: float,
        rate: float,
        leverage: float,
        reduceOnly: bool = False,
        time_in_force: str = "GTC",
    ) -> dict:
        res = super().create_order(
            pair=pair,
            ordertype=ordertype,
            side=side,
            amount=amount,
            rate=rate,
            leverage=leverage,
            reduceOnly=reduceOnly,
            time_in_force=time_in_force,
        )
        # Hyperliquid returns only the order-id.

        if not self._config["dry_run"]:
           res = self._api.fetch_order(res['id'], pair)
        return res
```

let me know what you think!

xmatthias replied:
well moving it to hyperliquid won't change the point - if the order is placed _but_ the `fetch_order` call doesn't receive the order yet (timing problem, sometimes seen during peak times) - freqtrade would "forget" this order - as you get an "order not found" error while placing an order - so from a freqtrade perspective - no order was ever placed - hence the trade is never stored (if it happens on the initial order).

---

With kucoin, the result from the exchange contains _only_ the id.
CCXT provides us with that - and all other information as None / undefined.

if hyperliquid returns actual values (`0`, not `None`) - then that should be fixed in ccxt - as that'd mean they're inventing values that are not there. Values not passed by the upstream API and without ability to derrive from other values (like - calculating cost as amount * price) - should be set to None / undefined.

The only thing ccxt does wrong on kucoin is the order status - which we simply override in the response.
If ccxt doesn't fix this - i think the better option is to "null" all values the api doesn't provide (pretty sure they do fix this though - as i think that's an undetected bug so far) - which will be everything except for the actual id i'd guess.

https://github.com/freqtrade/freqtrade/blob/1bf71b47df1ccdb8520524d55c6a54f62fd59a6c/freqtrade/exchange/kucoin.py#L65-L68

gaardiolor replied:
I think ccxt is ok, here's what ccxt returned:

```
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:13,663 - freqtrade.exchange.exchange - INFO - Order create_order(ETH/USDC:USDC, limit, buy, 0.0196, 2624.9, {'timeInForce': 'GTC'}
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:14,876 - freqtrade.exchange.exchange - INFO - API create_order: {'info': {'resting': {'oid': '42205402675'}}, 'id': '42205402675', 'clientOrderId': None, 'timestamp': None, 'datetime': None, 'lastTradeTimestamp': None, 'lastUpdateTimestamp': None, 'symbol': None, 'type': 'limit', 'timeInForce': None, 'postOnly': None, 'reduceOnly': None, 'side': None, 'price': None, 'triggerPrice': None, 'amount': None, 'cost': None, 'average': None, 'filled': None, 'remaining': None, 'status': 'open', 'fee': None, 'trades': [], 'fees': [], 'stopPrice': None, 'takeProfitPrice': None, 'stopLossPrice': None}
```

so amount is `None`.. Which for some reason seemed to result in the telegram message amount being 0 (because after doing the explicit fetch_order, I did not see that anymore)

```
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:15,795 - freqtrade.rpcrpc.rpc_manager - INFO - Sending rpc message: {'trade_id': 6, 'type': entry, 'buy_tag': 'timestamp_enter_long', 'enter_tag': 'timestamp_enter_long', 'exchange': 'Hyperliquid', 'pair': 'ETH/USDC:USDC', 'leverage': 2.0, 'direction': 'Long', 'limit': 2624.9, 'open_rate': 2624.9, 'order_type': 'limit', 'stake_amount': 25.790319685350003, 'stake_currency': 'USDC', 'base_currency': 'ETH', 'quote_currency': 'USDC', 'fiat_currency': 'USD', 'amount': 0.0, 'open_date': datetime.datetime(2024, 10, 16, 18, 5, 14, 879725, tzinfo=datetime.timezone.utc), 'current_rate': 2624.9, 'sub_trade': False}
```

Now looking at the code I don't really understand why that happened.. need to investigate it again

gaardiolor replied:
ok, so in freqtradebot.py `execute_entry()` 

* order['amount'] is `None`. 
* [here](https://github.com/freqtrade/freqtrade/blob/63dfa52ea85c2d8f9dcabba5c457f0dc7c75e0b4/freqtrade/freqtradebot.py#L1002) trade.amount is set to 0. 

in `_notify_enter()` [here](https://github.com/freqtrade/freqtrade/blob/63dfa52ea85c2d8f9dcabba5c457f0dc7c75e0b4/freqtrade/freqtradebot.py#L1216C13-L1216C95) it says `"amount": order.safe_amount_after_fee if fill else (order.amount or trade.amount),`

I logged that: `2024-11-03 22:05:15,452 - freqtrade.freqtradebot - INFO - order.safe_amount_after_fee: 0.0 fill: False order.amount: None trade.amount: 0.0`

So.. that explains why the telegram message is sent with amount 0. Not sure how it works for kucoin?

Instead of `self._api.fetch_order(order["id"], pair)`, I could overwrite amount None with the actual amount passed to create_order.. but since a lot of other values are None as well, I don't know if only overwriting the amount is a good/safe approach.

xmatthias replied:
it works the same way - and has just been fixed in develop (in a completely different way).

I don't see a reason to call `fetch_order` (or do any other shenanigans like setting random fields to things) - as freqtrade handles a response from create_order like `{'id': 'orderidasdf123'}` perfectly fine (proven by kucoin, actually).
Now the ccxt response won't look like this - but will have all attributes filled as "none" - and as long as that's the case - there's no problem.

gaardiolor replied:
ok, https://github.com/freqtrade/freqtrade/commit/5cb4a65cbd67b36e8b10ef34d999d62f2ad9ac8a I guess. If it's fixed there I can of course delete my workarounds. I'll test it.

gaardiolor replied:
seems to work! /status shows Amount: 0 for a few seconds though, not sure if that's something you want to fix. I'll revert my 'workaround' commits.

gaardiolor replied:
done https://github.com/freqtrade/freqtrade/pull/10839/commits/38eb4eed97bfc408e6324503d02e98d65c0f359e

xmatthias replied:
> seems to work! /status shows Amount: 0 for a few seconds though, not sure if that's something you want to fix. I'll revert my 'workaround' commits.

no that's an intentional change. 
You don't actually own anything until the order actually fills (think - limit order that never fills) - so the amount that was shown previously was wrong.

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -16,20 +20,143 @@ class Hyperliquid(Exchange):
+    ) -> int:
+        # Funding rate candles have a different limit
+        if candle_type == CandleType.FUNDING_RATE:
+            return 500
```
Have you got a source for this ? 
The docs don't seem to state what the limit for this is (?)

gaardiolor replied:
Well, I found it by testing:

```
    e = ccxt.hyperliquid()
    e.aiohttp_proxy = 'http://127.0.0.1:8080'
    ts = int(datetime(2023, 1, 1, 0, 0).timestamp() * 1000)
    history = await e.fetch_funding_rate_history('BTC/USDC:USDC', ts)
    print(len(history))
    await e.close()
```
`500`

That gives 500 candles. Seems to start when the market went live:
- first candle 1683849600048 Friday, May 12, 2023 12:00:00.048 AM
- last candle 1687687200201 Sunday, June 25, 2023 10:00:00.201 AM

Same when talking directly to the hyperliquid api, 

```
curl -s -X POST https://api.hyperliquid.xyz/info --json '{"type":"fundingHistory","coin":"BTC","interval":"5m","startTime":1672527600000}' | jq | grep coin | wc -l
```
`500`


In ccxt, when not providing a ts, it gives the 100 most recent candles. 

```
    e = ccxt.hyperliquid()
    e.aiohttp_proxy = 'http://127.0.0.1:8080'
    history = await e.fetch_funding_rate_history('BTC/USDC:USDC')
    print(len(history))
    print(history[0])
    print(history[-1])
    await e.close()
```

```
100
{'info': {'coin': 'BTC', 'fundingRate': '0.0000125', 'premium': '0.00015252', 'time': '1730124000111'}, 'symbol': 'BTC/USDC:USDC', 'fundingRate': 1.25e-05, 'timestamp': 1730124000111, 'datetime': '2024-10-28T14:00:00.111Z'}
{'info': {'coin': 'BTC', 'fundingRate': '0.0000125', 'premium': '0.00043811', 'time': '1730480400046'}, 'symbol': 'BTC/USDC:USDC', 'fundingRate': 1.25e-05, 'timestamp': 1730480400046, 'datetime': '2024-11-01T17:00:00.046Z'}
```

what ccxt does is calculating the timestamp 100 hours ago and passing that as the start timestamp ([here](https://github.com/ccxt/ccxt/blob/4a1c0c6c1756732ab7c09fc238f920302baa4df6/python/ccxt/async_support/hyperliquid.py#L1544)).

xmatthias replied:
which seems like _yet another_ bug ( :laughing: ) - as they're ignoring the limit parameter . . .

gaardiolor replied:
Hm sort of, they accept it, but to a max of 100 instead of 500. I've logged an issue https://github.com/ccxt/ccxt/issues/24144

gaardiolor replied:
fixed in ccxt 4.4.26

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -16,20 +20,143 @@ class Hyperliquid(Exchange):
+            leverage = int(leverage)
+            # Hyperliquid needs the parameter leverage.
+            # Don't use set_leverage(), as this sets margin back to cross
+            self.set_margin_mode(pair, self.margin_mode, params={"leverage": leverage})
```
Not necessarily a problem - but either of the 2 could be used, each with the "other" parameter set via params 

for set_leverage: 
https://github.com/ccxt/ccxt/blob/4b5cfd7cd8db24f86f79ac3dd69d39a34cba5c18/ts/src/hyperliquid.ts#L2378

gaardiolor replied:
ok, not sure though if `_set_leverage()` (missed the leading '_' in my comment, I'll fix that) can be used though, it doesn't [have ](https://github.com/freqtrade/freqtrade/blob/1bf71b47df1ccdb8520524d55c6a54f62fd59a6c/freqtrade/exchange/exchange.py#L3294) a params parameter

```
    def _set_leverage(
        self,
        leverage: float,
        pair: Optional[str] = None,
        accept_fail: bool = False,
    ):
```

set_margin_mode() does [have ](https://github.com/freqtrade/freqtrade/blob/1bf71b47df1ccdb8520524d55c6a54f62fd59a6c/freqtrade/exchange/exchange.py#L3344) it:
```
    def set_margin_mode(
        self,
        pair: str,
        margin_mode: MarginMode,
        accept_fail: bool = False,
        params: Optional[dict] = None,
    ):
```

the `set_leverage()` in ccxt does [have](https://github.com/ccxt/ccxt/blob/4a1c0c6c1756732ab7c09fc238f920302baa4df6/python/ccxt/async_support/hyperliquid.py#L2210) it - but I'm using the `_set_leverage()` of the base freqtrade exchange class, not the one of ccxt.

xmatthias replied:
Well as said - it's not a problem - just a comment - i think ccxt will use the same exact endpoint under the hood anyway.

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -16,20 +20,143 @@ class Hyperliquid(Exchange):
+        maintenance_margin_required = position_value / max_leverage / 2
+
+        # Docs: margin_available (isolated) = isolated_margin - maintenance_margin_required
+        margin_available = isolated_margin - maintenance_margin_required
```
I'd recommend to use `wallet_balance` here (instead of isolated_margin) - that'd auto-prepare this function for cross mode.

This seems to be the only difference between isolated and cross (unless i miss-read the docs)- the difference is taken in `liquidation_price.py` - which passes either stake_amount (what you're using here) - or the overall account balance (currently not yet correct - but the fix will eventually happen there).

gaardiolor replied:
ok, I'll fix that

gaardiolor replied:
(fixed)

xmatthias commented on tests/exchange/test_hyperliquid.py:
```diff
@@ -0,0 +1,319 @@
+            "contracts": 0.00067,
+            "collateral": 14.979079,
+            "leverage": 3.0,
+            "liquidationPrice": 45236.52992613,
```
Just to understand this right - these are positions you actually opened, - which we can then use as calculation reference, right?

xmatthias replied:
nevermind - i think https://github.com/freqtrade/freqtrade/pull/10839#discussion_r1821476969 actually covers this.

xmatthias reviewed (commented):
Overall, looks good (under the assumption that it works - which i didn't test yet, but plan to do sometime over the weekend).

I'd want to add an explicit section to the exchanges.md page - which either links to how to set up the API - or explains how to do that (or most likely, a combination of both) - as i don't think that'll be clear to the average user (it's a new setting in the exchange structure, a new way on how to generate that key, ...)

Commit:
_lev_prep: fixed typo. dry_run_liquidation_price: use wallet_balance instead of stake_amount

gaardiolor commented:
> I'd want to add an explicit section to the exchanges.md page - which either links to how to set up the API - or explains how to do that (or most likely, a combination of both) - as i don't think that'll be clear to the average user (it's a new setting in the exchange structure, a new way on how to generate that key, ...)

I have drafted something up here: https://github.com/gaardiolor/freqtrade/blob/hyperliquid-docs/docs/exchanges.md . You can use it as a starting point, or I can create another pull request, or I can include it in this one.. let me know.

xmatthias commented:
Best add it here i'd say.

from a content perspective - if hyperliquid has some docs about how to deposit (essentially the whole part after how to generate the keys) - i'd rather link there - mostly in hopes that they'll have more uptodate info going forward (i doubt we'll notice if that part get's outdated and also remember to update this) :laughing:

gaardiolor committed:
Update exchanges.md - add Hyperliquid

gaardiolor commented:
> from a content perspective - if hyperliquid has some docs about how to deposit (essentially the whole part after how to generate the keys) - i'd rather link there - mostly in hopes that they'll have more uptodate info going forward (i doubt we'll notice if that part get's outdated and also remember to update this) 😆

Yeah I agree, I fixed that and added some warnings / best practices, let me know if you want to keep those.

Commit:
Reverted changes related to create_order_has_all_data

gaardiolor committed:
Update exchanges.md - typo

gaardiolor committed:
Update exchanges.md - add hyperliquid stoploss_on_exchange note

gaardiolor committed:
Update stoploss.md - add hyperliquid

gaardiolor commented:
Added stoploss_on_exchange to the documentation. 

ccxt/hyperliquid also supports market stoploss orders. However, that still needs a [price](https://github.com/ccxt/ccxt/blob/9ae45ed0b665dcfe70d21b0c442273c1fa19bf10/python/ccxt/async_support/hyperliquid.py#L1132). I don't think freqtrade supports this currently, and how much sense it makes to support it. Might be the same as a limit combined with stoploss_on_exchange_limit_ratio = 0.95 ? 

Only added the 'limit' stoploss_on_exchange type to the docs.

xmatthias committed:
docs: fix hyperliquid docs formatting

xmatthias commented:
> Might be the same as a limit combined with stoploss_on_exchange_limit_ratio = 0.95 ?

well that's for sure correct.
Leave it as is for now - adding this in a generic way is probably a 2 line fix - though not something i'd see as necessary for a DEX (i think market orders are WAY more dangerous on a DEX than on a CEX).
Also given by the fact that ccxt applies a 5% max slippage, it's not "actual" limit orders - but limit orders with a price offset of 5% "into" the other orderbook.

xmatthias committed:
tests: improve hyperliquid tests

gaardiolor commented:
ok!

What do you think, anything left to do before this one can be merged? Let me know if I can do anything.

xmatthias committed:
tests: add hyperliquid lev_prep test

xmatthias commented:
> What do you think, anything left to do before this one can be merged? Let me know if I can do anything.

at the moment - no.
the code itself looks good - i'd just like to do a couple of tests against the actual exchange - which will probably take me a few days as my schedule RN is pretty full.

xmatthias committed:
chore: fix bad import

xmatthias commented on freqtrade/exchange/hyperliquid.py:
```diff
@@ -16,20 +20,142 @@ class Hyperliquid(Exchange):
+        "l2_limit_range": [20],
+        "trades_has_history": False,
+        "tickers_have_bid_ask": False,
+        "stoploss_on_exchange": True,
```
This ain't true for spot though.


i got `ccxt.base.errors.ExchangeError: hyperliquid {"status":"ok","response":{"type":"order","data":{"statuses":[{"error":"Placing trigger orders is not enabled on spot trading yet."}]}}}
` - which essentially is an exchange level error stating that "we don't support spot stoploss orders".

Also, market orders (at least on spot, didn't get to test futures yet) don't work without further configuration.
It's an easy fix (see my latest commits) - but shows that more testing is necessary.

xmatthias commented on docs/stoploss.md:
```diff
@@ -36,6 +36,7 @@ The Order-type will be ignored if only one mode is available.
 | Gate     | limit |
 | Okx      | limit |
 | Kucoin   | stop-limit, stop-market|
+| Hyperliquid   | limit |
```
```suggestion
| Hyperliquid (futures only)   | limit |
```

This is only really true for futures markets - spot markets don't support stoploss orders (see other comment).
(might need additional edits in the exchange file)

xmatthias committed:
feat(hyperliquid): fix a few settings to make spot work properly

xmatthias commented on docs/exchanges.md:
```diff
@@ -295,6 +295,53 @@ It's therefore required to pass the UID as well.
+
+walletAddress must be in hex format: `0x<40 hex characters>`, and can be easily copied from your wallet.
+
+privateKey also must be in hex format: `0x<64 hex characters>`. 
```
are you sure this is true / necessary? 
I copy/pasted both walletAddress and PrivateKey directly out of metamask - and it did work fine (private key did NOT start with 0x).

Is there some documentation (ccxt or hyperliquid) to highlight that this would actually be necessary / improves performance or security ... ? 

I'd rather not recommend people install random dependencies - not considering the problem of pip package poisoning / typo squatting, which is especially prevalent in the crypto space.

gaardiolor replied:
ah you're right, 0x is not necessary. ccxt strips the leading 0x part by taking only the last 64 characters (example: `signature = self.ecdsa(hash[-64:], privateKey[-64:], 'secp256k1', None)` ). I'll update that. I'll also add that it can be exported from metamask. Not sure if all wallets support that, for some you might need the mnemonic and re-generate the private key like in my example. I'll add another warning, let me know if you want to totally delete it.

gaardiolor replied:
Hm, [here](https://github.com/ccxt/ccxt/blob/58074e6d03b96f0658ffb93dbce1db9d2f035bde/python/ccxt/base/exchange.py#L227) ccxt mentions it should be prefixed with 0x. It works without in this case, at this time, but maybe better to state that it should be prefixed with 0x to increase the chance that it will keep working in the future :)

xmatthias replied:
that works for me (0x won't hurt) - though i'm not sure the "convert from mnemonic" is really necessary - that seems to complicate things - it's more a "get a private key" topic - some wallets may not give you the mnemonic phrase (didn't test the ones hyperliquid supprots to be fair) - so the instruction may turn out to be rather missleading ...

gaardiolor replied:
ok, I will delete it

gaardiolor replied:
(done)

xmatthias reviewed (commented):
Spot markets do have a "strict" and "all" distinction.

Do we have a way to make this distinction, too? (non-strict markets seem to have essentially no volume?).

---


Also, volume pairlist yields an odd ordering (spot - futures seems fine) - are we sure that volume isn't in "coin" unit (that'd be a flag to switch).
While PURR was at the top - LQNA was next in line, with CABAL below  - which doesn't seem to make sense from an ordering perspective (that's based on "regular" VolumePairlist config).
Switching to "advanced" mode (specify timeframe and lookback period) - this seems to "readjust" itself - which is odd and might point to an exchange (or ccxt) bug in the "fetch_tickers" endpoint.

![image](https://github.com/user-attachments/assets/27b98a26-ad57-48fe-9475-078fc5550dda)
![image](https://github.com/user-attachments/assets/31961222-d248-4ed5-a3a9-60ada4eab39d)

gaardiolor committed:
Update docs/stoploss.md - clarify that stoploss_on_exchange is only supported for hyperliquid futures

Co-authored-by: Matthias <xmatthias@outlook.com>

xmatthias committed:
Merge branch 'develop' into pr/gaardiolor/10839

xmatthias committed:
chore: adopt hyperopt class to new ruff formatting

gaardiolor committed:
Update exchanges.md - further clarify private key best practices

gaardiolor committed:
Update exchanges.md - less details on the private key conversion

xmatthias committed:
docs: fix spelling

gaardiolor commented:
> Spot markets do have a "strict" and "all" distinction. Do we have a way to make this distinction, too? (non-strict markets seem to have essentially no volume?).

I don't think so, from an API standpoint it soesn't seem to make a difference.

> might point to an exchange (or ccxt) bug in the "fetch_tickers" endpoint.

Nice find! I've investigated it, it's a bug in ccxt. For some reason Hyperliquid spot tickers' endpoint is totally different (and less straightforward) compared to their futures api endpoint. Futures looks good.  

I'll log a ccxt issue.

xmatthias commented:
> > Spot markets do have a "strict" and "all" distinction. Do we have a way to make this distinction, too? (non-strict markets seem to have essentially no volume?).
> 
> I don't think so, from an API standpoint it soesn't seem to make a difference.


That's a shame - maybe should be reported to the devs ... as that's something that'd really be useful (the "non-strict" pairs have essentially no volume / candles).


> I'll log a ccxt issue.

:laughing:

gaardiolor commented:
> That's a shame - maybe should be reported to the devs ... as that's something that'd really be useful (the "non-strict" pairs have essentially no volume / candles).

I've done some searching on discord to figure out what makes a coin fall in the 'strict' category. Consensus that it seems to be loosely related to how long the coin has been around, market cap and volume. The Hyperliquid team also made some comments pointing in that direction. But, wouldn't be surprised if there's a fee involved as well. 

The full rules have not been publicly disclosed. I'm not sure how CCXT would be able to expose which coins are on the strict list and which are not.. I think VolumePairList is a better way to get the markets with sufficient volume - The 'strict' coin with the lowest volume (OMNIX) only had $4627, so I think VolumePairList is more reliable. 

Issue number 18 logged, I'm feeling this is one of the last :) https://github.com/ccxt/ccxt/issues/24240

gaardiolor commented:
https://github.com/ccxt/ccxt/issues/24240 is fixed, I've tested it, it's working now. Spot ticker data is accurate. 

It should be in ccxt 4.4.31 (hasn't been released yet)

xmatthias committed:
fix: hyperliquid market orders don't exist and always require price

xmatthias committed:
chore: reorder hyperliquid docs, add note about order types

xmatthias committed:
chore: add hyperliquid to supported exchange list

xmatthias committed:
chore: add Hyperliquid as supported exchange to docs

xmatthias commented:
> It should be in ccxt 4.4.31 (hasn't been released yet)

It still hasn't. I'm waiting for this as i'd like to test this for spot (current ordering is rather confusing) ... after that, this will be merged.

> The full rules have not been publicly disclosed. I'm not sure how CCXT would be able to expose which coins are on the strict list and which are not.

I think you missed my point - this wasn't about ccxt adding this (at this point, anyway)
i can see from the "info" section of the markets that the information isn't available in the API (that's what i meant).

It's not really a problem (from a bot perspective) to not know by which rules a pair get's "strict" status or not (a person might want to know this - in reality, a person will want to filter by this though).

What we would want to know is which pairs are in which group however - which i think hyperliquid should expose as part of the API endpoint - but they currently don't.

gaardiolor commented:
> It still hasn't. I'm waiting for this as i'd like to test this for spot (current ordering is rather confusing) ... after that, this will be merged.

ok, cool! 4.4.31 has just been released :)

PS: I've asked if they can expose which markets are on strict. But I since the conditions on which they determine which market ends up on the strict list seem to be a bit arbitrary, (volume/market cap does not seem the only aspect), I think VolumePairList might be more reliable

xmatthias committed:
chore: bump ccxt to 4.4.31

gaardiolor commented:
Hmm.. the 8h candle is new in ccxt ‘timeframes’, before it was 6h (which is not supported by hyperliquid). Strange that it fails now instead of before, will investigate later today

xmatthias reviewed (approved):
Great enhancement :100: 

The test failure is not related to hyperliquid (the same test (gate.io) fails on develop, too).

As such - LGTM :+1:

gaardiolor commented:
That's great. Learned some new things, was nice working with you @xmatthias ! Might do another dex in the future :)
